### PR TITLE
Type annotations for planemo.bioblend, planemo.git, and planemo.cwl.run

### PR DESCRIPTION
**Pull request by adRn-s**

### CODE REVIEW

**Review by nsoranzo — COMMENTED** [review]
Can you also run black again please?

**Comment by nsoranzo** [comment]
Ping @mr-c

File: planemo/tools.py [hunk 1]
```diff
@@ -34,7 +48,15 @@ def uris_to_paths(ctx, uris):
...
 
-def yield_tool_sources_on_paths(ctx, paths, recursive=False, yield_load_errors=True, exclude_deprecated=False):
+def yield_tool_sources_on_paths(
+    ctx: "PlanemoCliContext",
+    paths: Union[List[str], Tuple[str], Tuple[str, str]],
+    recursive: bool = False,
+    yield_load_errors: bool = True,
+    exclude_deprecated: bool = False,
+) -> Iterator[
+    Union[Tuple[str, "XmlToolSource"], Tuple[str, object], Tuple[str, "CwlToolSource"], Tuple[str, "YamlToolSource"]]
```
[the remark this answers is not in the document]
Comment (adRn-s) [reply planemo/tools.py:58]: Ah, thanks! I will look into this

File: planemo/tools.py [hunk 2]
```diff
@@ -3,22 +3,34 @@
...
+from typing import (
+    Any,
+    Iterator,
+    List,
+    Tuple,
+    TYPE_CHECKING,
+    Union,
+)
 
 from galaxy.tool_util import loader_directory
```
Comment (nsoranzo) [inline planemo/tools.py:15]:
```suggestion
from galaxy.tool_util import loader_directory
from galaxy.tool_util.loader_directory import (
    is_tool_load_error,
    TOOL_LOAD_ERROR,
)
```

File: planemo/tools.py [hunk 3]
```diff
@@ -3,22 +3,34 @@
...
 
 from planemo.io import (
     error,
     info,
 )
 
+if TYPE_CHECKING:
+    from planemo.cli import PlanemoCliContext
+
 is_tool_load_error = loader_directory.is_tool_load_error
```
Comment (nsoranzo) [inline planemo/tools.py:27]:
```suggestion
```

File: planemo/tools.py [hunk 4]
```diff
@@ -3,22 +3,34 @@
 import os
 import sys
 import traceback
+from typing import (
+    Any,
+    Iterator,
```
Comment (nsoranzo) [inline planemo/tools.py:8]:
```suggestion
    Iterable,
    Iterator,
```

File: planemo/tools.py [hunk 5]
```diff
@@ -34,7 +46,13 @@ def uris_to_paths(ctx, uris):
...
 
 
-def yield_tool_sources_on_paths(ctx, paths, recursive=False, yield_load_errors=True, exclude_deprecated=False):
+def yield_tool_sources_on_paths(
+    ctx: "PlanemoCliContext",
+    paths: Union[List[str], Tuple[str], Tuple[str, str]],
+    recursive: bool = False,
+    yield_load_errors: bool = True,
+    exclude_deprecated: bool = False,
+) -> Iterator[Tuple[str, "ToolSource"]]:
```
Comment (nsoranzo) [inline planemo/tools.py:55]:
```suggestion
) -> Iterator[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR]]]:
```

File: planemo/tools.py [hunk 6]
```diff
@@ -43,7 +61,9 @@ def yield_tool_sources_on_paths(ctx, paths, recursive=False, yield_load_errors=T
             yield (tool_path, tool_source)
 
 
-def yield_tool_sources(ctx, path, recursive=False, yield_load_errors=True):
+def yield_tool_sources(
+    ctx: "PlanemoCliContext", path: str, recursive: bool = False, yield_load_errors: bool = True
+) -> Iterator[Tuple[str, "ToolSource"]]:
```
Comment (nsoranzo) [inline planemo/tools.py:66]:
```suggestion
) -> Iterator[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR]]]:
```

File: planemo/tools.py [hunk 7]
```diff
@@ -63,7 +83,9 @@ def yield_tool_sources(ctx, path, recursive=False, yield_load_errors=True):
         yield (tool_path, tool_source)
 
 
-def load_tool_sources_from_path(path, recursive, register_load_errors=False):
+def load_tool_sources_from_path(
+    path: str, recursive: bool, register_load_errors: bool = False
+) -> List[Tuple[str, "ToolSource"]]:
```
Comment (nsoranzo) [inline planemo/tools.py:88]:
```suggestion
) -> List[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR]]]:
```

File: planemo/tools.py [hunk 8]
```diff
@@ -3,22 +3,38 @@
...
     info,
 )
 
-is_tool_load_error = loader_directory.is_tool_load_error
+if TYPE_CHECKING:
+    from planemo.cli import PlanemoCliContext
+
 SKIP_XML_MESSAGE = "Skipping XML file - does not appear to be a tool %s."
 SHED_FILES = ["tool_dependencies.xml", "repository_dependencies.xml"]
 LOAD_ERROR_MESSAGE = "Error loading tool with path %s"
```
Comment (nsoranzo) [inline planemo/tools.py:34]:
```suggestion
LOAD_ERROR_MESSAGE = "Error loading tool with path %s"
TOOL_LOAD_ERROR_type = type(TOOL_LOAD_ERROR)
```

File: planemo/tools.py [hunk 9]
```diff
@@ -34,7 +50,13 @@ def uris_to_paths(ctx, uris):
...
 
 
-def yield_tool_sources_on_paths(ctx, paths, recursive=False, yield_load_errors=True, exclude_deprecated=False):
+def yield_tool_sources_on_paths(
+    ctx: "PlanemoCliContext",
+    paths: Iterable[str],
+    recursive: bool = False,
+    yield_load_errors: bool = True,
+    exclude_deprecated: bool = False,
+) -> Iterator[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR]]]:
```
Comment (nsoranzo) [inline planemo/tools.py:59]:
```suggestion
) -> Iterator[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR_type]]]:
```

File: planemo/tools.py [hunk 10]
```diff
@@ -43,7 +65,9 @@ def yield_tool_sources_on_paths(ctx, paths, recursive=False, yield_load_errors=T
             yield (tool_path, tool_source)
 
 
-def yield_tool_sources(ctx, path, recursive=False, yield_load_errors=True):
+def yield_tool_sources(
+    ctx: "PlanemoCliContext", path: str, recursive: bool = False, yield_load_errors: bool = True
+) -> Iterator[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR]]]:
```
Comment (nsoranzo) [inline planemo/tools.py:70]:
```suggestion
) -> Iterator[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR_type]]]:
```

File: planemo/tools.py [hunk 11]
```diff
@@ -63,7 +87,9 @@ def yield_tool_sources(ctx, path, recursive=False, yield_load_errors=True):
         yield (tool_path, tool_source)
 
 
-def load_tool_sources_from_path(path, recursive, register_load_errors=False):
+def load_tool_sources_from_path(
+    path: str, recursive: bool, register_load_errors: bool = False
+) -> List[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR]]]:
```
Comment (nsoranzo) [inline planemo/tools.py:92]:
```suggestion
) -> List[Tuple[str, Union[ToolSource, TOOL_LOAD_ERROR_type]]]:
```

File: planemo/cwl/run.py [hunk 12]
```diff
@@ -28,29 +38,34 @@
 class CwlToolRunResponse(SuccessfulRunResponse):
     """Describe the resut of a cwltool invocation."""
 
-    def __init__(self, runnable, log, outputs=None):
-        self._runnable = runnable
+    def __init__(
+        self,
+        log: str,
+        outputs: Optional[Dict[str, Union[Dict[str, Union[str, int]], int]]] = None,
```
Comment (jmchilton) [inline planemo/cwl/run.py:44]: I wrote this code and don't understand that outputs type declaration - (Like why does that dictionary only container integers and strings) - my intuition says this should be ```Optional[Dict[str, Any]]``` instead. Or at least an ``Any`` on that inner ``Union``.

I'm happy to merge as is though and wait for a problem with this.

Comment (nsoranzo) [reply planemo/cwl/run.py]: @jmchilton Agreed, sometimes MonkeyType tries too hard! Fixed in a rebase.

**Pull request merged**